olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java

```diff
@@ -263,6 +274,151 @@ public int hashCode() {
+        if(aDerefOpt.isPresent() && bDerefOpt.isPresent()) {
+            // both are references
+            boolean eq = innerStructuralEquals(a, b, aSimple.getValue(), bSimple.getValue());
+            if(!eq) {
+                logger.info(String.format("Property key: %s, a does not structurally equal b", propName));
+            }
+            return eq;
+        } else if ((!aDerefOpt.isPresent()) && (!bDerefOpt.isPresent())) {
+            // both are not references
+            boolean valueMatch = aSimple.equals(bSimple);
+            if(!valueMatch) {
+                logger.info(String.format("Property key: %s, a.value: %s, b.value: %s",
+                        propName, aSimple.getValue(), bSimple.getValue()));
+            }
+            return valueMatch;
+        } else {
+            // mismatch between reference and non-reference
+            if(aDerefOpt.isPresent()) {
+                logger.info(String.format("Property key: %s a (%s) is a reference while b (%s) is a value", propName, aSimple.getValue(), bSimple.getValue()));
+            } else {
+                logger.info(String.format("Property key: %s a (%s) is a value while b (%s) is a reference", propName, aSimple.getValue(), bSimple.getValue()));
+            }
+            return false;
+        }
+    }
+
+    /**
+     * See {@link #structuralEquals(List, List, String, String)} for description of behavior.
+     */
+    private static boolean innerStructuralEquals(Map<String, ConfigurationData> a, Map<String, ConfigurationData> b, String aName, String bName) {
+        Optional<ConfigurationData> aRootOpt = Optional.ofNullable(a.get(aName));
+        Optional<ConfigurationData> bRootOpt = Optional.ofNullable(b.get(bName));
+        if((!aRootOpt.isPresent()) || (!bRootOpt.isPresent())) {
+            if(!aRootOpt.isPresent()) {
+                logger.info(String.format("%s is not found", aName));
+            } else {
+                logger.info(String.format("%s is not found", bName));
+            }
+            return false;
+        } else if(aRootOpt.get().equals(bRootOpt.get())) {
+            logger.info(String.format("%s and %s are .equals()", aName, bName));
+            return true;
+        } else {
+            ConfigurationData aRoot = aRootOpt.get();
+            ConfigurationData bRoot = bRootOpt.get();
+
+            Set<String> propNames = new HashSet<>(aRoot.getProperties().keySet());
+            propNames.addAll(bRoot.getProperties().keySet());
+
+            boolean typesMatch = aRoot.getClassName().equals(bRoot.getClassName());
+            if(!typesMatch) {
+                logger.info(String.format("type mismatch, a.class: %s b.class: %s", aRoot.getClassName(), bRoot.getClassName()));
+            }
+
+            return typesMatch &&
+                    propNames.stream().allMatch(propName -> {
+                        Optional<Property> aPropOpt = Optional.ofNullable(aRoot.getProperties().get(propName));
```

Comment:
Can we extract this lambda into a static method? It's hard to figure out what it's capturing and that level of complexity in a lambda body worries me (e.g. it's harder to debug). If possible could you recursively do that so the nested lambdas are also static methods? There are too many return statements for too many functions.